core/src/main/java/org/elasticsearch/cluster/service/InternalClusterService.java

```diff
@@ -413,6 +442,9 @@ public String source() {
         }
 
         assert batchResult.executionResults != null;
+        assert batchResult.executionResults.size() == toExecute.size();
+        final ClusterStateTaskExecutor.BatchResult<T> finalBatchResult = batchResult;
+        assert toExecute.stream().map(updateTask -> updateTask.task).allMatch(finalBatchResult.executionResults::containsKey);
```

Comment:
For the task-specific assertion message to be useful, tasks will need to override `Object#toString` lest a useless error message will be displayed. However, I agree that the machinery should be there to enable the possibility of more useful error messages. I pushed 02354fc491c2767448b21150b6ed0e58dfefb322.